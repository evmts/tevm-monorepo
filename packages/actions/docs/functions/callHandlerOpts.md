[**@tevm/actions**](../README.md) • **Docs**

***

[@tevm/actions](../globals.md) / callHandlerOpts

# Function: callHandlerOpts()

> **callHandlerOpts**(`client`, `params`): `Promise`\<`object` \| `object`\>

## Parameters

• **client**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

• **params**: [`CallParams`](../type-aliases/CallParams.md)\<`boolean`\>

## Returns

`Promise`\<`object` \| `object`\>

## Defined in

[packages/actions/src/Call/callHandlerOpts.js:19](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/Call/callHandlerOpts.js#L19)
