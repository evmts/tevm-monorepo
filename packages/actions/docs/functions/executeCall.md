[**@tevm/actions**](../README.md)

***

[@tevm/actions](../globals.md) / executeCall

# Function: executeCall()

> **executeCall**(`client`, `evmInput`, `params`, `events`?): `Promise`\<[`ExecuteCallResult`](../type-aliases/ExecuteCallResult.md) & `object` \| \{ `errors`: \[[`ExecuteCallError`](../type-aliases/ExecuteCallError.md)\]; \}\>

Defined in: [packages/actions/src/Call/executeCall.js:29](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/Call/executeCall.js#L29)

executeCall encapsalates the internal logic of running a call in the EVM

## Parameters

### client

`TevmNode`\<`"fork"` \| `"normal"`, \{\}\>

### evmInput

`EVMRunCallOpts`

### params

[`CallParams`](../type-aliases/CallParams.md)\<`boolean`\>

### events?

[`CallEvents`](../type-aliases/CallEvents.md)

Optional event handlers for EVM execution

## Returns

`Promise`\<[`ExecuteCallResult`](../type-aliases/ExecuteCallResult.md) & `object` \| \{ `errors`: \[[`ExecuteCallError`](../type-aliases/ExecuteCallError.md)\]; \}\>

## Throws

returns errors as values
