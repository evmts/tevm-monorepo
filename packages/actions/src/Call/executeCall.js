import { runTx } from '@tevm/vm'
import { bytesToHex, hexToBytes, numberToHex } from 'viem'
import { evmInputToImpersonatedTx } from '../internal/evmInputToImpersonatedTx.js'
import { runCallWithTrace } from '../internal/runCallWithTrace.js'
import { handleRunTxError } from './handleEvmError.js'
import { createAddress } from '@tevm/address'

/**
 * The error returned by executeCall
 * @internal
 * @typedef {import('./handleEvmError.js').HandleRunTxError} ExecuteCallError
 */

/**
 * The return value of executeCall
 * @internal
 * @typedef {{runTxResult: import("@tevm/vm").RunTxResult, trace: import('../debug/DebugResult.js').DebugTraceCallResult | undefined, accessList: undefined | Map<string, Set<string>>}} ExecuteCallResult
 */

/**
 * executeCall encapsalates the internal logic of running a call in the EVM
 * @internal
 * @param {import('@tevm/node').TevmNode} client
 * @param {import("@tevm/evm").EvmRunCallOpts} evmInput
 * @param {import('./CallParams.js').CallParams} params
 * @returns {Promise<(ExecuteCallResult & {errors?: [ExecuteCallError]}) | {errors: [ExecuteCallError]}>}
 * @throws {never} returns errors as values
 */
export const executeCall = async (client, evmInput, params) => {
	/**
	 * @type {import('../debug/DebugResult.js').DebugTraceCallResult | undefined}
	 */
	let trace = undefined
	/**
	 * evm returns an access list without the 0x prefix
	 * @type {Map<string, Set<string>> | undefined}
	 */
	let accessList = undefined
	const vm = await client.getVm()
		// as an optimization we want to attempt to fetch all storage up front
		; (async () => {
			if (!client.forkTransport) return
			const blockTag = await client.getVm()
				.then(vm => vm.blockchain.blocksByTag.get('forked'))
				.then(block => block !== undefined ? numberToHex(block.header.number) : 'latest')
			// TODO this is wrong we should use forkTransport.send but it doesn't exist yet
			/**
			 * @type {import('../eth/EthJsonRpcResponse.js').EthCreateAccessListJsonRpcResponse}
			 */
			const accessListRes = await client.forkTransport?.request({
				method: 'eth_createAccessList',
				params: [
					{
						to: evmInput.to,
						gas: evmInput.gasLimit,
						// gasPrice: evmInput.gasPrice, not necessary to set this
						value: evmInput.value,
						data: evmInput.data,
					},
					blockTag,
				],
			})
			if (accessListRes.error) {
				client.logger.error('Unable to get an access list', accessListRes.error)
			}
			for (const contract of accessListRes.result?.accessList ?? []) {
				for (const storageKey of contract.storageKeys) {
					const getStorage = () => vm.stateManager._baseState.caches.storage.get(createAddress(contract.address), hexToBytes(storageKey, { size: 32 }))
					if (getStorage()?.length) {
						return
					}
					/**
					 * @type {import('../eth/EthJsonRpcResponse.js').EthCreateAccessListJsonRpcResponse}
					 */
					client.forkTransport.request({
						method: 'eth_getStorageAt',
						params: [contract.address, storageKey, blockTag]
					}).then(async storage => {
						const vm = await client.getVm()
						// THIS MUST BE SYNC TO INITIALLY SET STORAGE OR RACE CONDITIONS WILL EXIST!!!
						// This can be cleaned up with a state lock later
						if (getStorage()?.length) {
							return
						}
						vm.stateManager._baseState.caches.storage
							.put(
								createAddress(contract.address),
								hexToBytes(storageKey, { size: 32 }),
								hexToBytes(/** @type {import('viem').Hex}*/(storage))
							)
					})
				}
			}
		})()
	try {
		const tx = await evmInputToImpersonatedTx({
			...client,
			getVm: () => Promise.resolve(vm),
		})(evmInput, params.maxFeePerGas, params.maxPriorityFeePerGas)
		if (params.createTrace) {
			// this trace will be filled in when the tx runs
			trace = await runCallWithTrace(vm, client.logger, evmInput, true).then(({ trace }) => trace)
		}
		const runTxResult = await runTx(vm)({
			reportAccessList: params.createAccessList ?? false,
			reportPreimages: params.createAccessList ?? false,
			skipHardForkValidation: true,
			skipBlockGasLimitValidation: true,
			// we currently set the nonce ourselves user can't set it
			skipNonce: true,
			// we must skipBalance for now because we have no clue what the gasLimit should be so this initial run we set it to block maximum
			skipBalance: true,
			...(evmInput.block !== undefined ? { block: /** @type any*/ (evmInput.block) } : {}),
			tx,
		})

		if (trace) {
			trace.gas = runTxResult.execResult.executionGasUsed
			trace.failed = false
			trace.returnValue = bytesToHex(runTxResult.execResult.returnValue)
		}

		client.logger.debug(
			{
				returnValue: bytesToHex(runTxResult.execResult.returnValue),
				exceptionError: runTxResult.execResult.exceptionError,
				executionGasUsed: runTxResult.execResult.executionGasUsed,
			},
			'callHandler: runCall result',
		)
		if (params.createAccessList && runTxResult.accessList !== undefined) {
			accessList = new Map(
				runTxResult.accessList.map((item) => {
					return [item.address, new Set(item.storageKeys)]
				}),
			)
		}

		return {
			runTxResult,
			accessList,
			trace,
			...(runTxResult.execResult.exceptionError !== undefined
				? { errors: [handleRunTxError(runTxResult.execResult.exceptionError)] }
				: {}),
		}
	} catch (e) {
		return {
			trace,
			accessList,
			errors: [handleRunTxError(e)],
		}
	}
}
