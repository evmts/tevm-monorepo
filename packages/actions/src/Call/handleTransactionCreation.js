import { createTransaction } from '../CreateTransaction/createTransaction.js'
import { handleAutomining } from './handleAutomining.js'
import { shouldCreateTransaction } from './shouldCreateTransaction.js'

/**
 * Handles the creation of a transaction based on the call parameters and execution result.
 *
 * @param {import('@tevm/node').TevmNode} client - The TEVM base client instance.
 * @param {import('./CallParams.js').CallParams} params - The call parameters.
 * @param {import('./executeCall.js').ExecuteCallResult} executedCall - The result of the executed call.
 * @param {import('@tevm/evm').EvmRunCallOpts} evmInput - The EVM input parameters.
 * @returns {Promise<{hash: import('@tevm/utils').Hex | undefined, errors?: never} | {hash?: never, errors: Array<import('./TevmCallError.js').TevmCallError>}>} A promise that resolves to the transaction hash or undefined.
 * @throws {never} Returns errors as values
 */
export const handleTransactionCreation = async (client, params, executedCall, evmInput) => {
	/**
	 * @type {import('@tevm/utils').Hex | undefined}
	 */
	let txHash = undefined
	/**
	 * @type {Array<import('./TevmCallError.js').TevmCallError>}
	 */
	const errors = []
	if (shouldCreateTransaction(params, executedCall.runTxResult)) {
		const txRes = await createTransaction(client)({
			throwOnFail: false,
			evmOutput: executedCall.runTxResult,
			evmInput,
			maxPriorityFeePerGas: params.maxPriorityFeePerGas,
			maxFeePerGas: params.maxFeePerGas,
		})
		txHash = 'txHash' in txRes ? txRes.txHash : undefined
		const miningRes = (await handleAutomining(client, txHash)) ?? {}
		const { errors: txCreationErrors } = /** @type {{errors: Array<any>}}*/ (txRes)
		errors.push(...(miningRes.errors ?? []), ...(txCreationErrors ?? []))
	}
	if (errors.length > 0) {
		return {
			errors,
		}
	}
	return { hash: txHash }
}
