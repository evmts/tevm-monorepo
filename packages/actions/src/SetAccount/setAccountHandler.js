import { createAddress } from '@tevm/address'
import { AccountNotFoundError, InternalError } from '@tevm/errors'
import { EthjsAccount } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateSetAccountParams } from './validateSetAccountParams.js'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./SetAccountHandlerType.js').SetAccountHandler}
 */
export const setAccountHandler =
	(client, options = {}) =>
	async (params) => {
		const { throwOnFail = options.throwOnFail ?? true } = params
		/**
		 * @type {Array<import('../SetAccount/TevmSetAccountError.js').TevmSetAccountError>}
		 */
		const errors = validateSetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}

		const address = createAddress(params.address)

		/**
		 * @type {Array<Promise<any>>}
		 */
		const promises = []
		try {
			const vm = await client.getVm()

			// check if account exists
			const account = await getAccountHandler(client)({ ...params, throwOnFail: false })
			if (account.errors?.length && !(account.errors[0] instanceof AccountNotFoundError)) {
				client.logger.error('there was an unexpected error getting account', account.errors)
				throw account.errors.length > 1 ? new AggregateError(account.errors) : account.errors[0]
			}

			promises.push(
				vm.stateManager.putAccount(
					address,
					new EthjsAccount(
						params.nonce ?? account?.nonce,
						params.balance ?? account?.nonce,
						(params.storageRoot && hexToBytes(params.storageRoot)) ??
							(account?.storageRoot !== undefined && account?.storageRoot !== '0x'
								? hexToBytes(account.storageRoot)
								: undefined),
						(params.deployedBytecode && hexToBytes(keccak256(params.deployedBytecode))) ??
							(account?.deployedBytecode !== undefined ? hexToBytes(keccak256(account.deployedBytecode)) : undefined),
					),
				),
			)
			if (params.deployedBytecode) {
				promises.push(vm.stateManager.putContractCode(address, hexToBytes(params.deployedBytecode)))
			}
			// state clears state first stateDiff does not
			if (params.state) {
				await vm.stateManager.clearContractStorage(address)
			}
			const state = params.state ?? params.stateDiff
			if (state) {
				for (const [key, value] of Object.entries(state)) {
					promises.push(
						vm.stateManager.putContractStorage(
							address,
							hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (key), { size: 32 }),
							hexToBytes(value),
						),
					)
				}
			}
			const results = await Promise.allSettled(promises)
			for (const result of results) {
				if (result.status === 'rejected') {
					errors.push(new InternalError('Unable to put storage', { cause: result.reason }))
				}
			}

			if (errors.length > 0) {
				return maybeThrowOnFail(throwOnFail, { errors })
			}
			await vm.stateManager.checkpoint()
			await vm.stateManager.commit(false)

			return {}
		} catch (e) {
			errors.push(new InternalError('Unexpected error setting account', { cause: e }))
			return maybeThrowOnFail(throwOnFail, { errors })
		}
	}
