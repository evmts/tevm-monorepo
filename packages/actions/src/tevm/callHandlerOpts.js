import { EthjsAddress } from '@tevm/utils'
import { hexToBytes } from '@tevm/utils'

/**
 * Parses user provided params into ethereumjs options to pass into the EVM
 * @param {import('@tevm/actions-types').CallParams} params
 * @returns {Parameters<import('@tevm/evm').Evm['runCall']>[0]}
 */
export const callHandlerOpts = (params) => {
	/**
	 * @type {Parameters<import('@tevm/evm').Evm['runCall']>[0]}
	 */
	const opts = {}

	if (params.to) {
		opts.to = EthjsAddress.fromString(params.to)
	}
	if (params.data) {
		opts.data = hexToBytes(params.data)
	}
	if (params.salt) {
		opts.salt = hexToBytes(params.salt)
	}
	if (params.depth) {
		opts.depth = params.depth
	}
	if (params.blobVersionedHashes) {
		opts.blobVersionedHashes = params.blobVersionedHashes.map((hash) =>
			hexToBytes(hash),
		)
	}
	if (params.selfdestruct) {
		opts.selfdestruct = params.selfdestruct
	}
	if (params.skipBalance) {
		opts.skipBalance = Boolean(params.skipBalance)
	}
	if (params.gasRefund) {
		opts.gasRefund = BigInt(params.gasRefund)
	}
	if (params.gasPrice) {
		opts.gasPrice = BigInt(params.gasPrice)
	}
	if (params.value) {
		opts.value = BigInt(params.value)
	}
	const caller = params.caller || params.from
	if (caller) {
		opts.caller = EthjsAddress.fromString(caller)
	}
	const origin = params.origin || params.from
	if (origin) {
		opts.origin = EthjsAddress.fromString(origin)
	}
	if (params.gas) {
		opts.gasLimit = BigInt(params.gas)
	}

	return opts
}
