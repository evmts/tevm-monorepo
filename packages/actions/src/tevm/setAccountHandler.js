import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAccount, EthjsAddress } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { validateSetAccountParams } from '@tevm/zod'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').SetAccountHandler}
 */
export const setAccountHandler = (client, options = {}) => async (params) => {
	const { throwOnFail = options.throwOnFail ?? true } = params
	/**
	 * @type {Array<import('@tevm/errors').SetAccountError>}
	 */
	const errors = validateSetAccountParams(params)
	if (errors.length > 0) {
		return maybeThrowOnFail(throwOnFail, { errors })
	}

	const address = new EthjsAddress(hexToBytes(params.address))
	try {
		await client.vm.stateManager.putAccount(
			address,
			new EthjsAccount(
				params.nonce,
				params.balance,
				params.storageRoot && hexToBytes(params.storageRoot),
				params.deployedBytecode &&
					hexToBytes(keccak256(params.deployedBytecode)),
			),
		)
		if (params.deployedBytecode) {
			await client.vm.stateManager.putContractCode(
				address,
				hexToBytes(params.deployedBytecode),
			)
		}
		await client.vm.stateManager.checkpoint()
		await client.vm.stateManager.commit()
		// TODO offer way of setting contract storage with evm.stateManager.putContractStorage
		return {}
	} catch (e) {
		errors.push(
			createError(
				'UnexpectedError',
				typeof e === 'string'
					? e
					: e instanceof Error
					? e.message
					: 'unknown error',
			),
		)
		return maybeThrowOnFail(throwOnFail, { errors })
	}
}
