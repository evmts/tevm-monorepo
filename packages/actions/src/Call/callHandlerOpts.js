import { createAddress } from '@tevm/address'
import { InvalidBlockError, InvalidParamsError, UnknownBlockError } from '@tevm/errors'
import { hexToBytes } from '@tevm/utils'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'

/**
 * @internal
 * @typedef {UnknownBlockError | UnknownBlockError | InvalidParamsError} CallHandlerOptsError
 */

/**
 * @internal
 * Parses user provided params into ethereumjs options to pass into the EVM
 * @param {import('@tevm/node').TevmNode} client
 * @param {import('./CallParams.js').CallParams} params
 * @returns {Promise<{data: Parameters<import('@tevm/evm').Evm['runCall']>[0], errors?: never} | {data?: never, errors: Array<CallHandlerOptsError>}>}
 * @throws { never } Returns all errors as values
 */
export const callHandlerOpts = async (client, params) => {
	/**
	 * @type {Parameters<import('@tevm/evm').Evm['runCall']>[0]}
	 */
	const opts = {}
	const vm = await client.getVm()

	// TODO need better error handling here
	const block = await (async () => {
		try {
			if (params.blockTag === undefined) {
				return vm.blockchain.blocksByTag.get('latest')
			}
			if (typeof params.blockTag === 'bigint') {
				return await vm.blockchain.getBlock(params.blockTag)
			}
			if (typeof params.blockTag === 'string' && params.blockTag.startsWith('0x')) {
				return await vm.blockchain.getBlock(hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (params.blockTag)))
			}
			// TODO support all these and resolve all of them both vs fork and non fork
			if (
				params.blockTag === 'latest' ||
				params.blockTag === 'safe' ||
				params.blockTag === 'pending' ||
				params.blockTag === 'earliest' ||
				params.blockTag === 'finalized'
			) {
				return vm.blockchain.blocksByTag.get(/** */ params.blockTag)
			}
			return new InvalidBlockError(`Unknown blocktag ${params.blockTag}`)
		} catch (e) {
			return new UnknownBlockError(e instanceof Error ? e.message : `Unable to find block ${params.blockTag}`)
		}
	})()
	if (block instanceof UnknownBlockError || block instanceof InvalidBlockError || block === undefined) {
		return { errors: [block ?? new UnknownBlockError(`Unable to find block ${params.blockTag}`)] }
	}

	client.logger.debug({ block: block.header }, 'Using block')

	opts.block = block

	// handle block overrides
	if (params.blockOverrideSet) {
		client.logger.debug(params.blockOverrideSet, 'callHandlerOpts: Detected a block override set')
		// TODO this is a known bug we need to implement better support for block tags
		// We are purposefully ignoring this until the block creation is implemented
		const { header } = await vm.blockchain.getCanonicalHeadBlock()
		opts.block = {
			...opts.block,
			header: {
				coinbase:
					params.blockOverrideSet.coinbase !== undefined
						? createAddress(params.blockOverrideSet.coinbase)
						: header.coinbase,
				number: params.blockOverrideSet.number !== undefined ? BigInt(params.blockOverrideSet.number) : header.number,
				difficulty: header.difficulty,
				prevRandao: header.prevRandao,
				gasLimit:
					params.blockOverrideSet.gasLimit !== undefined ? BigInt(params.blockOverrideSet.gasLimit) : header.gasLimit,
				timestamp: params.blockOverrideSet.time !== undefined ? BigInt(params.blockOverrideSet.time) : header.timestamp,
				baseFeePerGas:
					params.blockOverrideSet.baseFee !== undefined
						? BigInt(params.blockOverrideSet.baseFee)
						: header.baseFeePerGas ?? BigInt(0),
				cliqueSigner() {
					return header.cliqueSigner()
				},
				getBlobGasPrice() {
					if (params.blockOverrideSet?.blobBaseFee !== undefined) {
						return BigInt(params.blockOverrideSet.blobBaseFee)
					}
					return header.getBlobGasPrice()
				},
			},
		}
	}

	// handle state overrides
	if (params.stateOverrideSet) {
		client.logger.debug(params.stateOverrideSet, 'callHandlerOpts: Detected a stateOverrideSet')
		for (const [address, state] of Object.entries(params.stateOverrideSet)) {
			const res = await setAccountHandler(client)({
				address: /** @type import('@tevm/utils').Address*/ (address),
				...(state.nonce !== undefined ? { nonce: state.nonce } : {}),
				...(state.balance !== undefined ? { balance: state.balance } : {}),
				...(state.code !== undefined ? { deployedBytecode: state.code } : {}),
				...(state.state !== undefined ? { state: state.state } : {}),
				...(state.stateDiff !== undefined ? { stateDiff: state.stateDiff } : {}),
				throwOnFail: false,
			})
			if (res.errors?.length) {
				return {
					errors: [
						new InvalidParamsError('Invalid state override', {
							cause: /** @type {Error} */ (res.errors.length === 1 ? res.errors[0] : new AggregateError(res.errors)),
						}),
					],
				}
			}
		}
	}

	if (params.to) {
		opts.to = createAddress(params.to)
	}
	if (params.data) {
		opts.data = hexToBytes(params.data)
	}
	if (params.salt) {
		opts.salt = hexToBytes(params.salt)
	}
	if (params.depth) {
		opts.depth = params.depth
	}
	if (params.blobVersionedHashes) {
		opts.blobVersionedHashes = params.blobVersionedHashes.map((hash) => hexToBytes(hash))
	}
	if (params.selfdestruct) {
		opts.selfdestruct = params.selfdestruct
	}
	if (params.gasRefund) {
		opts.gasRefund = BigInt(params.gasRefund)
	}
	if (params.gasPrice) {
		opts.gasPrice = BigInt(params.gasPrice)
	}
	if (params.value) {
		opts.value = BigInt(params.value)
	}
	const caller =
		params.caller ||
		params.from ||
		params.origin ||
		(params.createTransaction ? '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266' : `0x${'00'.repeat(20)}`)
	if (caller) {
		opts.caller = createAddress(caller)
	}
	const origin =
		params.origin ||
		params.from ||
		(params.createTransaction ? '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266' : `0x${'00'.repeat(20)}`)
	if (origin) {
		if (params.skipBalance !== undefined) {
			opts.skipBalance = Boolean(params.skipBalance)
		} else {
			opts.skipBalance = caller === `0x${'00'.repeat(20)}` && (params.createTransaction ?? false) === false
		}
		opts.origin = createAddress(origin)
	}
	if (params.gas) {
		opts.gasLimit = BigInt(params.gas)
	}

	if (params.createTransaction && opts.block !== (await vm.blockchain.getCanonicalHeadBlock())) {
		throw new Error('Creating transactions on past blocks is not currently supported')
	}

	return { data: opts }
}
// 36,45-46,84,90
