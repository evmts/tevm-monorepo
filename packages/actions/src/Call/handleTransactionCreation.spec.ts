import { createAddress } from '@tevm/address'
import { createTevmNode } from '@tevm/node'
import { TestERC20 } from '@tevm/test-utils'
import { encodeFunctionData } from 'viem'
import { describe, expect, it } from 'vitest'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'
import type { CallParams } from './CallParams.js'
import { callHandlerOpts } from './callHandlerOpts.js'
import { executeCall } from './executeCall.js'
import { handleTransactionCreation } from './handleTransactionCreation.js'

const contract = TestERC20.withAddress(createAddress(420420420420420).toString())

describe(handleTransactionCreation.name, async () => {
	it('should handle transaction creation', async () => {
		const client = createTevmNode()

		await setAccountHandler(client)({
			address: contract.address,
			deployedBytecode: contract.deployedBytecode,
		})

		const params: CallParams = {
			data: encodeFunctionData(contract.read.balanceOf(createAddress(25).toString())),
			to: contract.address,
			gas: 16784800n,
			createTransaction: 'on-success',
		}

		const vm = await client.getVm().then((vm) => vm.deepCopy())

		const { data: evmInput, errors: callHandlerOptsErrors } = await callHandlerOpts(client, params)

		expect(callHandlerOptsErrors).toBeUndefined()
		if (evmInput === undefined) throw new Error('evmInput should be undefined')

		const executeCallResult = await executeCall({ ...client, getVm: async () => vm }, evmInput, params)

		if ('errors' in executeCallResult) throw new Error('executeCallResult.errors should be undefined')

		const { hash, errors } = await handleTransactionCreation(client, params, executeCallResult, evmInput)

		expect(errors).toBeUndefined()
		expect(hash).toBeDefined()
		expect(hash).toMatchSnapshot()
	})

	it('should do nothing if createTransaction is false', async () => {
		const client = createTevmNode()

		await setAccountHandler(client)({
			address: contract.address,
			deployedBytecode: contract.deployedBytecode,
		})

		const params: CallParams = {
			data: encodeFunctionData(contract.read.balanceOf(createAddress(25).toString())),
			to: contract.address,
			gas: 16784800n,
			createTransaction: false,
		}

		const vm = await client.getVm().then((vm) => vm.deepCopy())

		const { data: evmInput, errors: callHandlerOptsErrors } = await callHandlerOpts(client, params)

		expect(callHandlerOptsErrors).toBeUndefined()
		if (evmInput === undefined) throw new Error('evmInput should be undefined')

		const executeCallResult = await executeCall({ ...client, getVm: async () => vm }, evmInput, params)

		if ('errors' in executeCallResult) throw new Error('executeCallResult.errors should be undefined')

		const { hash, errors } = await handleTransactionCreation(client, params, executeCallResult, evmInput)
		expect(errors).toBeUndefined()
		expect(hash).toBeUndefined()
	})

	it('should handle createTransaction having errors', async () => {
		const client = createTevmNode()

		const params: CallParams = {
			to: createAddress(0).toString(),
			createTransaction: true,
			value: 420n,
			from: createAddress(420420420).toString(),
		}

		const vm = await client.getVm().then((vm) => vm.deepCopy())

		const { data: evmInput, errors: callHandlerOptsErrors } = await callHandlerOpts(client, {
			...params,
			skipBalance: true,
		})

		expect(callHandlerOptsErrors).toBeUndefined()
		if (evmInput === undefined) throw new Error('evmInput should be undefined')

		const executeCallResult = await executeCall({ ...client, getVm: async () => vm }, evmInput, params)

		if ('errors' in executeCallResult) throw new Error('executeCallResult.errors should be undefined')

		const { hash, errors } = await handleTransactionCreation(client, params, executeCallResult, evmInput)
		expect(errors).toBeDefined()
		expect(hash).toBeUndefined()
		expect(errors).toMatchSnapshot()
	})
})
