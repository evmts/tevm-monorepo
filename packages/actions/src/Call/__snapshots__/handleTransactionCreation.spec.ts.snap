// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleTransactionCreation > should handle createTransaction having errors 1`] = `
[
  {
    "_tag": "NoBalanceError",
    "message": "Attempting to create a transaction with an uninitialized account with no balance",
    "name": "NoBalanceError",
  },
]
`;

exports[`handleTransactionCreation > should handle transaction creation 1`] = `"0x1a3e54004085fad191398087b59e055926584d1493696257aff39dd04d5a8fff"`;
